src/aws.s3/amazons3_errors.bal

```diff
@@ -0,0 +1,527 @@
+        | MethodNotAllowedError | MissingAttachmentError | MissingContentLengthError | MissingRequestBodyError
+        | MissingSecurityElementError | MissingSecurityHeaderError | NoLoggingStatusForKeyError | NoSuchBucketError
+        | NoSuchBucketPolicyError | NoSuchKeyError | NoSuchLifecycleConfigurationError | NoSuchUploadError
+        | NoSuchVersionError | NotImplementedError | NotSignedUpError | OperationAbortedError | PermanentRedirectError
+        | PreconditionFailedError | RedirectError | RestoreAlreadyInProgressError | RequestIsNotMultiPartContentError
+        | RequestTimeoutError | RequestTimeTooSkewedError | RequestTorrentOfBucketError
+        | ServerSideEncryptionConfigurationNotFoundError | ServiceUnavailableError | SignatureDoesNotMatchError
+        | SlowDownError| TemporaryRedirectError | TokenRefreshRequiredError | TooManyBucketsError | UnexpectedContentError
+        | UnresolvableGrantByEmailAddressError | UserKeyMustBeSpecifiedError;
+
+# Identifies unknown server error.
+public const UNKNOWN_SERVER_ERROR = "{ballerinax/aws.s3}UnknownServerError";
+# Represents unknown server error.
+public type UnknownServerError error<UNKNOWN_SERVER_ERROR, ErrorDetail>;
+
+# Ballerina AmazonS3 Union Errors
+public type ConnectorError ServerError|ClientError|StringUtilError;
+
+public type ServerError BucketOperationError|UnknownServerError;
+
+# Error messages.
+const string UNKNOWN_SERVER_ERROR_MSG = "Unknown Amazon S3 server error occured.";
+const string EMPTY_VALUES_FOR_CREDENTIALS_ERROR_MSG = "Empty values set for accessKeyId or secretAccessKey credential";
+const string CLIENT_CREDENTIALS_VERIFICATION_ERROR_MSG = "Error occured while verifying client credentials.";
+const string DATE_STRING_GENERATION_ERROR_MSG = "Error occured while generating date strings.";
+const string CANONICAL_URI_GENERATION_ERROR_MSG = "Error occured hwile generating canonical URI.";
+const string CANONICAL_QUERY_STRING_GENERATION_ERROR_MSG = "Error occured while generating canonical query string.";
+const string STRING_MANUPULATION_ERROR_MSG = "Error occured during string manipulation.";
+const string XML_EXTRACTION_ERROR_MSG = "Error occurred while accessing the XML payload from the http response.";
+const string API_INVOCATION_ERROR_MSG = "Error occurred while invoking the AmazonS3 API while ";
+const string BINARY_CONTENT_EXTRACTION_ERROR_MSG = "Error occured while accessing binary content from the http response";
+
+# Prepare the `error` as Amazons3 specific `ClientError`.
+#
+# + errorType - the error type.
+# + message - the error message.
+# + cause - the `error` instance.
+# + return - prepared `s3:ClientError` instance.
+public function prepareError(ClientErrorType errorType, string message, error? cause = ()) returns ClientError {
+    if (cause is error) {
+        error err = error(errorType, message = message, cause = cause);
+        return <ClientError> err;
+    } else {
+        error err = error(errorType, message = message);
```

Comment:
Can't we just use 
```suggestion
        return error(errorType, message = message);
```

here? AFAIR, we can now directly return error value. (Previously we had to cast this due to a compiler limitation). Please check whether we can directly return here, so that we can get rid of unnecessary variable creation and the casting.

Reply:
No. We have to still cast it here else the following compilation error occurs.

`error: ballerinax/aws.s3:0.11.0::amazons3_errors.bal:525:16: cannot infer type of the error from '(ballerinax/aws.s3:0.11.0:ClientCredentialsVerificationError|ballerinax/aws.s3:0.11.0:ClientConfigInitializationError|ballerinax/aws.s3:0.11.0:SignatureGenerationError|ballerinax/aws.s3:0.11.0:ApiInvocationError|ballerinax/aws.s3:0.11.0:HttpResponseHandlingError)'`

Reply:
Ah right then.